Simple SQL-joins with QueryBuilder

hllizi opened the pull request:
I added some basic support for inner joins to IHP's query builder module. It's now possible, for instance, to query the posts by author "Tom" as follows:

```
tomPosts <- query @Post
        |> innerJoin @User (#authorId, #id)
        |> filterWhereJoinedTable @User (#name, "Tom")
        |> fetch
```
This generates and excutes a query that obtains a set of posts which can be selected by reference to columns from the table of users, where the tables are joined `ON posts.author_id = users.id`. A short summary of the implemented features:

- A table can be joined on a column from a previous table using `innerJoinThirdTable`. E.g. the following code could be used to query all posts by users from department 5.
```
query @Posts 
        |> innerJoin @User (#authorId, #id) 
        |> innerJoinThirdTable @Department @User (#id, #departmentId) 
        |> filterWhereJoinedTable @Department (#number, 5)
``` 
- This also opens up the possibility of using the query builder module for many-to-many relations. The following code could be used to get all posts with the tag _haskell_.
```
query @Posts
        |> innerJoin @Tagging (#id, #postId)
        |> innerJoinThirdTable @Tag @Tagging (#id, #tagId)
        |> filterWhereJoinedTable @Tag (#tagText, "haskell")
        |> fetch
```
- Type safety is maintained by adding all joined types to a type-level list (type class IsJoined) and checking that the table has been joined where necessary. E.g. uses of `filterWhereJoinedTable @Model` will only compile if `@Model` had been joined to the input before.
- Backward compatibility should be fully guaranteed. I have tested this with a small (and the only, so far) project available to me and experienced no problems. More extensive testing might be in order, though.

mpscholten reviewed (commented):
Thanks a lot, great work 💯 

Commented a couple minor changes :) 

Also: maybe we could have some parahraphs about joins on https://ihp.digitallyinduced.com/Guide/relationships.html as well. This would increase aware about this feature as a lot of people don't check the haddock.

hllizi commented on IHP/Fetch.hs:
```diff
@@ -48,34 +48,78 @@ class Fetchable fetchable model | fetchable -> model where
+
+
+{-# INLINE commonFetch #-}
+commonFetch :: forall model table q r.(HasQueryBuilder q r, model ~ GetModelByTableName table, KnownSymbol table, KnownSymbol (GetTableName model), PG.FromRow model, ?modelContext :: ModelContext) => q table -> IO [model]
```
Done.

hllizi commented on IHP/QueryBuilder.hs:
```diff
@@ -116,6 +131,35 @@ data OrderByClause =
     , orderByDirection :: !OrderByDirection }
     deriving (Show, Eq)
 
+data NoJoins
+data EmptyModelList
+data ConsModelList model models
```
Done. Also added a type class ModelList to tell the true list types apart from the type NoJoins. innerJoin and innerJoinThirdTable no use this type class to ascertain that joining is allowed.

hllizi commented:
OK, as far as I can see, everything is resolved. I hope I'll find the time soon to add something to the guide, to.

hllizi commented:
It isn't quite clear to me yet what causes the failing test.

mpscholten commented:
Thanks :) Looks good so far.

I think the test fail is caused by the line `| JoinQueryBuilder       { queryBuilder :: !(QueryBuilder table), join :: Join}` in the `QueryBuilder.hs`.

The `join` field leads to an `join` getter function being exported from the `IHP.QueryBuilder` module. The `IHP.Prelude` also exports a [join](https://hackage.haskell.org/package/base-4.15.0.0/docs/Control-Monad.html#v:join). This leads to a conflict: The `IHP.ControllerPrelude` modules re-exports the `IHP.QueryBuilder` and also the `IHP.Prelude`.

A possible solution could be to add a `hiding` to the `IHP.ControllerPrelude` like `import IHP.QueryBuilder hiding (join)`.

hllizi commented:
OK, obvious, in retrospect. m(

Fixed it by renaming join to joinData. It's only for internal use and employed in, I think, just three places. It doesn't seem to matter much what it's named.

I also implemented filterWhereNotJoinedTable, which was still missing, and a corresponding test case.

mpscholten reviewed (approved):
Looks good now :) We should add some docs to the guide soon, but i think we can merge already 👍
